File: bst/lb-04-inorder-predecessor-successor.py
# Predecessor -> 1 left + extreme right
# Successor   -> 1 right + extreme left
# while going left maintain succ for leaf and only one child node.
# while going right maintain pred for ...

def pred(root):
    # 1 left + extreme right
    root = root.left
    while root and root.right:
        root = right.right
    return root


def succ(root):
    # 1 right + extreme left
    root = root.right
    while root and root.left:
        root = right.left
    return root


# in range pred and succ
def pred_succ_in_range(root, key):
    if not root:
        return
    if root.key == key:
        if root.left:
            pre = pred(root)
        if root.right:
            suc = succ(root)
        return
    elif root.key > key:
        # lies in left
        suc = root
        pred_succ_in_range(root.left, key)
    else:
        # lies in right
        pred = root
        pred_succ_in_range(root.right, key)

File: bst/cpp/lb-04-inorder-predecessor-successor.cpp

Node *pred(Node *root) {
    // 1 left then
    // extreme right
    root = root->left;
    while (root->right) {
        root = root->right;
    }
    return root;
}
Node *succ(Node *root) {
    // 1 right then
    // extreme left
    root = root->right;
    while (root->left) {
        root = root->left;
    }
    return root;
}
void find_pre_succ(Node *root, Node *pre, Node *suc, int key) {
    if (!root) return;
    if (root->key == key) {
        // if key is found the normally find the predecessor and successor.
        if (root->left) {
            pre = pred(root);
        }
        if (root->right) {
            suc = succ(root);
        }
        return;
    }
    // key exists in the right side of the tree
    if (root->key < key) {
        // maitain a predecessor as right child may not have left child
        // so root will be the predecessor.
        pre = root;
        find_pre_succ(root->right, pre, suc, key);
    } else {
        // key is less i.e in left side of tree
        // maintain the successor as left child may not have right child
        // so root will be the successor of child.
        suc = root;
        find_pre_succ(root->left, pre, suc, key);
    }
}